mlflow/langchain/databricks_dependencies.py

```diff
@@ -0,0 +1,185 @@
+from typing import Dict, Set
+
+_DATABRICKS_DEPENDENCY_KEY = "databricks_dependency"
+_DATABRICKS_VECTOR_SEARCH_INDEX_NAME_KEY = "databricks_vector_search_index_name"
+_DATABRICKS_VECTOR_SEARCH_ENDPOINT_NAME_KEY = "databricks_vector_search_endpoint_name"
+_DATABRICKS_EMBEDDINGS_ENDPOINT_NAME_KEY = "databricks_embeddings_endpoint_name"
+_DATABRICKS_LLM_ENDPOINT_NAME_KEY = "databricks_llm_endpoint_name"
+_DATABRICKS_CHAT_ENDPOINT_NAME_KEY = "databricks_chat_endpoint_name"
+
+
+def _assign_value_or_append_to_list(d, key, value):
+    if key in d:
+        if isinstance(d[key], list):
+            d[key].append(value)
+        else:
+            d[key] = [d[key], value]
+    else:
+        d[key] = value
+    return d
+
+
+def _extract_databricks_dependencies_from_retriever(retriever, dependency_dict) -> Dict[str, str]:
+    import langchain
+
+    if hasattr(retriever, "vectorstore") and hasattr(retriever.vectorstore, "embeddings"):
```

Comment:
could we use a helper function that uses `getattr(obj, instance, None)` and updates the dict within the helper to help to reduce the number of `hasattr` calls within this func?

Something like...

```python
def _try_append_dependency(vector_store_attr, dependency_key, attribute_name):
        attr = getattr(vector_store_attr, attribute_name, None)
        if attr is not None:
            _assign_value_or_append_to_list(dependency_dict, dependency_key, attr)

vectorstore = getattr(retriever, "vectorstore", None)
if vectorstore:
    embeddings = getattr(vectorstore, "embeddings", None)

    if isinstance(vectorstore, langchain.vectorstores.DatabricksVectorSearch):
        _try_append_dependency(vectorstore.index, _DATABRICKS_VECTOR_SEARCH_INDEX_NAME_KEY, "name")
        _try_append_dependency(vectorstore.index, _DATABRICKS_VECTOR_SEARCH_ENDPOINT_NAME_KEY, "endpoint_name")

    if isinstance(embeddings, langchain.embeddings.DatabricksEmbeddings):
        _try_append_dependency(embeddings, _DATABRICKS_EMBEDDINGS_ENDPOINT_NAME_KEY, "endpoint")
    elif callable(getattr(vectorstore, "_is_databricks_managed_embeddings", None)) and vectorstore._is_databricks_managed_embeddings():
        _assign_value_or_append_to_list(dependency_dict, _DATABRICKS_EMBEDDINGS_ENDPOINT_NAME_KEY, "_is_databricks_managed_embeddings")

return dependency_dict
```

Reply:
Seems like _try_append_dependency is not always needed.
If we create dependency_dict as collections.defaultdict(list), we can simplify _assign_value_or_append_to_list to dependency_dict[key].append(value).